src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Drivetrain.java

```diff
@@ -0,0 +1,355 @@
+                    MotorControllerFactory.createSparkMax(turnBackLeftPort, TemperatureLimit.NEO),
+                    MotorControllerFactory.createCANCoder(canCoderPortBL), 2,
+                    pitchSupplier, rollSupplier);
+            // Backward-Right
+            SwerveModule moduleBR = new SwerveModule(swerveConfig, SwerveModule.ModuleType.BR,
+                    MotorControllerFactory.createSparkMax(driveBackRightPort, TemperatureLimit.NEO),
+                    MotorControllerFactory.createSparkMax(turnBackRightPort, TemperatureLimit.NEO),
+                    MotorControllerFactory.createCANCoder(canCoderPortBR), 3,
+                    pitchSupplier, rollSupplier);
+            modules = new SwerveModule[] { moduleFL, moduleFR, moduleBL, moduleBR };
+        }
+        SmartDashboard.putNumber("kpTheta", thetaPIDController[0]);
+        odometry = new SwerveDrivePoseEstimator(kinematics, Rotation2d.fromDegrees(getHeading()), getModulePositions(), new Pose2d());
+        prevTime = Timer.getFPGATimestamp();
+    }
+
+    @Override
+    public void periodic() {
+        for (SwerveModule module : modules) module.periodic();
+
+        // Update the odometry with current heading and encoder position
+        odometry.update(Rotation2d.fromDegrees(getHeading()), getModulePositions());
+
+        // Update the odometry with limelight
+        if(lime != null && lime.hasTarget()) {
+            odometry.addVisionMeasurement(
+                lime.getTransform(limelightTransformForPoseEstimation).toPose2d(),
+                Timer.getFPGATimestamp() - lime.getNTEntry(limelightTransformForPoseEstimation.name().toLowerCase()).getDoubleArray(new double[7])[6] / 1000);
+        }
+
+        // SmartDashboard.putNumber("Odometry X",
+        // odometry.getPoseMeters().getTranslation().getX());
+        // SmartDashboard.putNumber("Odometry Y",
+        // odometry.getPoseMeters().getTranslation().getY());;
+        //SmartDashboard.putNumber("Pitch", gyro.getPitch());
+        //SmartDashboard.putNumber("Roll", gyro.getRoll());
+       // SmartDashboard.putNumber("Raw gyro angle", gyro.getAngle());
+        SmartDashboard.putNumber("Robot Heading", getHeading());
+        fieldOriented = SmartDashboard.getBoolean("Field Oriented", true);
+        // SmartDashboard.putNumber("Gyro Compass Heading", gyro.getCompassHeading());
+        // SmartDashboard.putNumber("Compass Offset", compassOffset);
+        // SmartDashboard.putBoolean("Current Magnetic Field Disturbance",
+        // gyro.isMagneticDisturbance());
+
+        // drive to point testing
+
+        // PID testing for rotations
+        thetaPIDController[0] = SmartDashboard.getNumber("kpTheta", thetaPIDController[0]);
+        double temp = gyro.getAngle();
+        double tempTime = Timer.getFPGATimestamp();
+        SmartDashboard.putNumber("Angular Vel", Units.degreesToRadians(temp - prevAngle) / (tempTime - prevTime));
```

Comment:
you can remove this. This was just me testing whether the drivetrain can drive at a constant angular velocity (which it didn't). There probably is a method in the kinematics class that does this.